plaidml/bridge/openvino/ops/proposal.cpp

```diff
@@ -0,0 +1,369 @@
+  anchor_buffer.copy_from(anchor_coordinates_vec.data());
+  auto anchor_offsets = edsl::Constant(anchor_buffer, "anchors_node");
+  anchor_offsets = edsl::cast(anchor_offsets, dtype);
+  auto anchors = anchor_offsets + center;
+
+  if (shift_anchors) {
+    anchors = anchors - base_size * 0.5f;
+  }
+  return anchors;
+}
+
+edsl::Tensor enumerate_proposals(edsl::Tensor class_probs,                //
+                                 edsl::Tensor class_logits,               //
+                                 edsl::Tensor anchors,                    //
+                                 const ngraph::op::ProposalAttrs& attrs,  //
+                                 const float img_H,                       //
+                                 const float img_W,                       //
+                                 const float scale_H,                     //
+                                 const float scale_W,                     //
+                                 const float coordinates_offset,          //
+                                 DType dtype) {
+  auto min_box_H = attrs.min_size * scale_H;
+  auto min_box_W = attrs.min_size * scale_W;
+
+  auto class_probs_shape = class_probs.compute_shape().sizes();
+  auto num_batches = class_probs_shape[0];
+  auto num_anchors = class_probs_shape[1] / 2;
+  auto feat_H = class_probs_shape[2];
+  auto feat_W = class_probs_shape[3];
+  auto num_proposals = feat_H * feat_W * num_anchors;
+
+  bool initial_clip = attrs.framework == "tensorflow";
+  bool swap_xy = attrs.framework == "tensorflow";
+
+  // used for gather index
+  edsl::Tensor idx_zero = edsl::index({edsl::TensorDim(1)}, 0);
+  edsl::Tensor idx_one = idx_zero + 1;
+  edsl::Tensor idx_two = idx_zero + 2;
+  edsl::Tensor idx_three = idx_zero + 3;
+
+  // anchors: [num_anchors, 4] -> [feat_H * feat_W * num_anchors, 4] = [num_proposals, 4])
+  anchors = op::tile(anchors, {static_cast<int>(feat_H * feat_W)});
+  edsl::Tensor anchor_wm = edsl::gather(anchors, idx_zero).axis(1);
+  edsl::Tensor anchor_hm = edsl::gather(anchors, idx_one).axis(1);
+  edsl::Tensor anchor_wp = edsl::gather(anchors, idx_two).axis(1);
+  edsl::Tensor anchor_hp = edsl::gather(anchors, idx_three).axis(1);
+
+  std::vector<float> feat_h_vec;
+  std::vector<float> feat_w_vec;
+  for (int64_t idx_feat_h = 0; idx_feat_h < feat_H; idx_feat_h++) {
+    for (int64_t idx_feat_w = 0; idx_feat_w < feat_W; idx_feat_w++) {
+      for (int64_t idx_anchor = 0; idx_anchor < num_anchors; idx_anchor++) {
+        feat_h_vec.push_back(static_cast<float>(idx_feat_h));
+        feat_w_vec.push_back(static_cast<float>(idx_feat_w));
+      }
+    }
+  }
+  TensorShape proposal_coord_shape(DType::FLOAT32, {num_proposals, 1});
+  auto feat_x_vec = swap_xy ? feat_h_vec : feat_w_vec;
+  auto feat_y_vec = swap_xy ? feat_w_vec : feat_h_vec;
+  Buffer feat_x_buffer(proposal_coord_shape);
+  Buffer feat_y_buffer(proposal_coord_shape);
+  feat_x_buffer.copy_from(feat_x_vec.data());
```

Comment:
Please use `index` instead of `copy_from` to generate the tensors.

Reply:
Done